### PR TITLE
Chapter 11: Account activation

### PR DESCRIPTION
**Pull request by stephanie1997-kek**

### CODE REVIEW

**Comment by stephanie1997-kek** [comment]
ready

File: .gitignore [hunk 1]
```diff
@@ -36,3 +36,6 @@ config/environments/*.local.yml
 /yarn-error.log
 yarn-debug.log*
 .yarn-integrity
+
+# Ignore application configuration
```
Comment (chungpv-1008) [inline .gitignore]: remove this line

File: app/assets/stylesheets/account_activations.scss [hunk 2]
```diff
@@ -0,0 +1,3 @@
+// Place all the styles related to the AccountActivations controller here.
```
Comment (chungpv-1008) [inline app/assets/stylesheets/account_activations.scss]: remove this file

File: app/views/user_mailer/account_activation.text.erb [hunk 3]
```diff
@@ -0,0 +1,3 @@
+<%= t ".greeting" %> <%= @user.name %>,
+<%= t ".introduction" %>
+<%= edit_account_activation_url(id: @user.activation_token, email: @user.email) %>
```
Comment (chungpv-1008) [inline app/views/user_mailer/account_activation.text.erb:3]:
```suggestion
<%= edit_account_activation_url id: @user.activation_token, email: @user.email %>
```

Comment (stephanie1997-kek) [reply app/views/user_mailer/account_activation.text.erb:3]: done

File: app/assets/stylesheets/account_activations.scss (hunk 2 above)
Comment (stephanie1997-kek) [reply app/assets/stylesheets/account_activations.scss]: done

File: .gitignore (hunk 1 above)
Comment (stephanie1997-kek) [reply .gitignore]: done

**Comment by stephanie1997-kek** [comment]
ready

File: app/controllers/account_activations_controller.rb [hunk 4]
```diff
@@ -0,0 +1,14 @@
+class AccountActivationsController < ApplicationController
+  def edit
+  	user = User.find_by email: params[:email]
+    if user && !user.activated? && user.authenticated? :activation, params[:id]
```
Comment (saobv-2223) [inline app/controllers/account_activations_controller.rb]:
```suggestion
  def edit
    user = User.find_by email: params[:email]
    if user && !user.activated? && user.authenticated? :activation, params[:id]
```

File: app/helpers/account_activations_helper.rb [hunk 5]
```diff
@@ -0,0 +1,2 @@
+module AccountActivationsHelper
+end
```
Comment (saobv-2223) [inline app/helpers/account_activations_helper.rb]: remove file that never used

File: app/mailers/user_mailer.rb [hunk 6]
```diff
@@ -0,0 +1,12 @@
+class UserMailer < ApplicationMailer
+
+  def account_activation
```
Comment (saobv-2223) [inline app/mailers/user_mailer.rb]:
```suggestion
  def account_activation
```

File: app/models/user.rb [hunk 7]
```diff
@@ -37,17 +39,31 @@ def remember
   end
 
   def authenticated? attribute, token
-    return false unless remember_digest
-    BCrypt::Password.new(remember_digest).is_password? remember_token
+    digest = send "#{attribute}_digest"
+    return false unless digest
+    BCrypt::Password.new(digest).is_password? token
```
Comment (saobv-2223) [inline app/models/user.rb]:
```suggestion
    return false unless digest
    
    BCrypt::Password.new(digest).is_password? token
```

File: app/models/user.rb [hunk 8]
```diff
@@ -37,17 +39,31 @@ def remember
...
+
+  def send_activation_email
+    UserMailer.account_activation(self).deliver_now
+  end
+
   private 
 
   def downcase_email
-    email.downcase!
+    self.email.downcase!
```
Comment (saobv-2223) [inline app/models/user.rb]:
```suggestion
    email.downcase!
```

File: config/application.yml [hunk 9]
```diff
@@ -0,0 +1,11 @@
+# Add configuration values here, as shown below.
+#
+# pusher_app_id: "2954"
```
Comment (saobv-2223) [inline config/application.yml]: please, ignore this file

Comment (stephanie1997-kek) [reply config/application.yml]: done

File: app/models/user.rb (hunk 8 above)
Comment (stephanie1997-kek) [reply app/models/user.rb]: done

File: app/models/user.rb (hunk 7 above)
Comment (stephanie1997-kek) [reply app/models/user.rb]: done

File: app/mailers/user_mailer.rb (hunk 6 above)
Comment (stephanie1997-kek) [reply app/mailers/user_mailer.rb]: done

File: app/helpers/account_activations_helper.rb (hunk 5 above)
Comment (stephanie1997-kek) [reply app/helpers/account_activations_helper.rb]: done

File: app/controllers/account_activations_controller.rb (hunk 4 above)
Comment (stephanie1997-kek) [reply app/controllers/account_activations_controller.rb]: done

**Comment by stephanie1997-kek** [comment]
ready

File: config/application.yml [hunk 10]
```diff
@@ -0,0 +1,11 @@
+# Add configuration values here, as shown below.
```
Comment (saobv-2223) [inline config/application.yml]: open .gitignore file and add config/application.yml
Then run git rm --cached config/application.yml to untrack this file. Finally, commit and push again

**Comment by stephanie1997-kek** [comment]
ready

**Review by saobv-2223 — COMMENTED** [review]
merge

**Pull request merged**